flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricStore.java

```diff
@@ -35,109 +46,111 @@
-				case "3":
-					if (components.length != 5) {
-						break;
-					}
-					job = jobs.get(components[1]);
+				case INFO_CATEGORY_TASK:
+					QueryScopeInfo.TaskQueryScopeInfo taskInfo = (QueryScopeInfo.TaskQueryScopeInfo) info;
+					job = jobs.get(taskInfo.jobID);
 					if (job == null) {
 						job = new JobMetricStore();
-						jobs.put(components[1], job);
+						jobs.put(taskInfo.jobID, job);
 					}
-					task = job.tasks.get(components[2]);
+					task = job.tasks.get(taskInfo.vertexID);
 					if (task == null) {
 						task = new TaskMetricStore();
-						job.tasks.put(components[2], task);
+						job.tasks.put(taskInfo.vertexID, task);
 					}
-					task.metrics.put(components[3] + "." + components[4], value);
+					/**
+					 * As the WebInterface task metric queries currently do not account for subtasks we don't 
+					 * divide by subtask and instead use the concatenation of subtask index and metric name as the name. 
+					 */
+					addMetric(task.metrics, taskInfo.subtaskIndex + "." + name, metric);
 					break;
-				/**
-				 * Operator metric
-				 * format: 4:<job_id>:<task_id>:<subtask_index>:<operator_name>:<user_scope>.<name>
-				 *
-				 * As the WebInterface does not account for operators (because it can't) we don't 
-				 * divide by operator and instead use the concatenation of subtask index, operator name and metric name 
-				 * as the name.
-				 */
-				case "4":
-					if (components.length != 6) {
-						break;
-					}
-					job = jobs.get(components[1]);
+				case INFO_CATEGORY_OPERATOR:
+					QueryScopeInfo.OperatorQueryScopeInfo operatorInfo = (QueryScopeInfo.OperatorQueryScopeInfo) info;
+					job = jobs.get(operatorInfo.jobID);
 					if (job == null) {
 						job = new JobMetricStore();
-						jobs.put(components[1], job);
+						jobs.put(operatorInfo.jobID, job);
 					}
-					task = job.tasks.get(components[2]);
+					task = job.tasks.get(operatorInfo.vertexID);
 					if (task == null) {
 						task = new TaskMetricStore();
-						job.tasks.put(components[2], task);
+						job.tasks.put(operatorInfo.vertexID, task);
 					}
-					task.metrics.put(components[3] + "." + components[4] + "." + components[5], value);
+					/**
+					 * As the WebInterface does not account for operators (because it can't) we don't 
+					 * divide by operator and instead use the concatenation of subtask index, operator name and metric name 
+					 * as the name.
+					 */
+					addMetric(task.metrics, operatorInfo.subtaskIndex + "." + operatorInfo.operatorName + "." + name, metric);
 					break;
 				default:
-					LOG.debug("Invalid metric name format: " + name);
+					LOG.debug("Invalid metric dump category: " + info.getCategory());
 			}
 		} catch (Exception e) {
-			LOG.debug("Malformed metric name format: " + name);
+			LOG.debug("Malformed metric dump.", e);
+		}
+	}
+
+	private void addMetric(Map<String, Object> target, String name, MetricDump metric) {
+		switch (metric.getCategory()) {
+			case METRIC_CATEGORY_COUNTER:
```

Comment:
The same question applies here as well. Why don't we simply check what type the metric instance is?